### PR TITLE
Batch user node state diagnostics and improve layout switching performance

### PR DESCRIPTION
**Pull request by shuhaowu**
**User-Facing Changes**

Improved layout switching performance when user scripts are present

**Description**

After user nodes are created, the code runs through a loop of `_setUserNodeDiagnostics` for each user script. These in turn calls React set state functions. Since set state functions are not batched in an async function below React 18 (and I see the latest branch is using React 17), this results in synchronous component updates, which in a complex layout can result in a 100ms latency per call. In a complex layout with many scripts (I have 9), this can take about ~2 seconds of run time (see attached screenshot in PR). Note that I used a previous build which also had `_setNodeMetadata` in that loop. This means with the current code and my layout, that would cause a ~1s lag.

Moving this loop into a batched update makes the block only costs 200ms, see below. With the current code without `setNodeMetadata`, this will only be 100ms.

Moving to React 18 should remove the need for this, but it would need to be confirmed.

image

image

### CODE REVIEW

File: packages/studio-base/src/players/UserNodePlayer/index.ts [hunk 1]
```diff
@@ -748,9 +749,11 @@ export default class UserNodePlayer implements Player {
       this._memoizedNodeDatatypes = nodeDatatypes;
     }
 
-    for (const nodeRegistration of state.nodeRegistrations) {
-      this._setUserNodeDiagnostics(nodeRegistration.nodeId, []);
-    }
+    ReactDOM.unstable_batchedUpdates(() => {
```
Comment (defunctzombie) [inline packages/studio-base/src/players/UserNodePlayer/index.ts:752]: Please copy some portion of your PR description here so we preserve the notes on performance and state setting for future readers.

**Review by defunctzombie — APPROVED** [review]
Once the comment is added and CLA signed we will merge.

**Review by snosenzo — APPROVED** [review]

**Comment by shuhaowu** [comment]
> Once the comment is added and CLA signed we will merge.

Both of these are done.

**Comment by shuhaowu** [comment]
Ah I had to add my work email here. Should be good now.

**Pull request merged**